os/fs/smartfs/smartfs_utils.c

```diff
@@ -1716,139 +1962,316 @@ int smartfs_countdirentries(struct smartfs_mountpt_s *fs, struct smartfs_entry_s
+		fdbg("Unable to write to sector %d while shrinking file, error = %d\n", readwrite.logsector, ret);
+		goto errout_with_filepos;
+	}
+#endif
+
+	nextsector = SMARTFS_NEXTSECTOR(chainheader);
+	chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
 
-		smartfs_setbuffer(&readwrite, nextsector, 0, sizeof(struct smartfs_chain_header_s), (uint8_t *)fs->fs_rwbuffer);
+	while (nextsector != SMARTFS_ERASEDSTATE_16BIT) {
+		smartfs_setbuffer(&readwrite, nextsector, 0, sizeof(struct smartfs_chain_header_s), (FAR uint8_t *)fs->fs_rwbuffer);
 		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
 		if (ret < 0) {
-			fdbg("Error reading sector %d header, ret : %d\n", readwrite.logsector, ret);
-			return ret;
+			fdbg("Unable to read sector %d, error = %d\n", readwrite.logsector, ret);
+			goto errout_with_filepos;
 		}
+		sector = SMARTFS_NEXTSECTOR(chainheader);
+		ret = FS_IOCTL(fs, BIOC_FREESECT, (unsigned long)nextsector);
+		if (ret < 0) {
+			fdbg("Unable to free sector %d to shrink file, error = %d\n", nextsector, ret);
+			goto errout_with_filepos;
+		}
+		nextsector = sector;
+	}
 
-		/* Get the next chained sector */
+errout_with_filepos:
+	sf->filepos = savefilepos;
+	return ret;
+}
 
-		sector = SMARTFS_NEXTSECTOR(header);
+/****************************************************************************
+ * Name: smartfs_extendfile
+ *
+ * Description:
+ *   Zero-extend the length of a regular file to 'length'.
+ *
+ ****************************************************************************/
 
-		/* If this is the 1st sector of the file, then just overwrite
-		 * the sector data with the erased state value.  The underlying
-		 * SMART block driver will detect this and release the old
-		 * sector and create a new one with the new (blank) data.
-		 */
+int smartfs_extendfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
+{
+	int ret;
+	uint16_t savefilepos;
+	uint16_t remaining;
+	size_t byteswritten;
+	uint16_t data_len;
+	struct smart_read_write_s readwrite;
+	char *buf;
 
-		if (nextsector == entry->firstsector) {
-#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	if (length == 0) {
+		return OK;
+	}
 
-			/* When we have a sector buffer in use, simply skip the first sector */
+	byteswritten = 0;
+	remaining = length;
+	savefilepos = sf->filepos;
 
-			nextsector = sector;
-			continue;
+	smartfs_seek_internal(fs, sf, 0, SEEK_END);
 
+	data_len = fs->fs_llformat.availbytes - sf->curroffset;
+	if (data_len > remaining) {
+		data_len = remaining;
+	}
+
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	smartfs_setbuffer(&readwrite, sf->currsector, 0, fs->fs_llformat.availbytes, (uint8_t *)sf->buffer);
+	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Error in reading sector %d, ret : %d\n", readwrite.logsector, ret);
+		goto errout_with_filepos;
+	}
+	/* Fill the current last sector till the end with '\0' */
+	memset(&sf->buffer[sf->curroffset], '\0', data_len);
```

Comment:
I think it is duplicated action, line 2167 will do this again, could you please verify it?

Reply:
Yes, optimizations possible